jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletContextResponse.java

```diff
@@ -278,6 +281,18 @@ public long getContentLength()
         return _contentLength;
     }
 
+    @Override
+    public void write(boolean last, ByteBuffer content, Callback callback)
+    {
+        _written += BufferUtil.length(content);
+        super.write(last, content, callback);
+    }
+
+    public long getContentBytesWritten()
+    {
+        return _written;
+    }
```

Comment:
> @gregw the test failure was due to welcome files bypassing the Servlet's OutputStream (using core classes directly), so that the check on the content-length was failing (as 0 bytes were written through the HttpOutput). Please re-review.

I'm not exactly how this method address this issue? The `getContentBytesWritten()` here looks like it will still be bypassed?  But fixed the tests... so I'm good... but an explanation would help.

Also, this method has the same name as `Response.getContentBytesWritten(Response)`, but returns a different value (i.e. pre compressed rather than post compressed).      Can you put some javadoc on both methods.   If you can think of a better name for this one, then change the name... maybe just `getWritten()` ?  Maybe make it package protected?

Reply:
@gregw I made the method package local, added javadocs, but also filed #11325 because the issue needs to be investigated further, but it is orthogonal to GzipHandler.

Reply:
The bypassing in question is done by the `ResourceService` used in the `DefaultServlet` when a static file is served and `DefaultServlet` notices there is no wrapping. In that case, `DefaultServlet` passes the `ServletContextResponse` directly to `ResourceService` instead of wrapping the `HttpServletRequest` as a core `Request` and passing that to `ResourceService`.

The problem this creates here is that `HttpOutput` is unaware of what `ResourceService` writes, and `getBytesWritten` returns how many bytes were written to the socket after `GZIPHandler` compressed them.

So `ServletChannel.handle()` had no reliable way to know how many bytes were written to check if `Content-Length` is correct, hence that extra bytes counter in `ServletContextResponse` was added so that `ServletChannel.handle()` can make that `Content-Length` check.

Reply:
I'm thinking that the Servlet stuff only needs to track bytes written enough to do the auto commit/flush/close required by the spec.   Perhaps we are doing too much due to tests left over from Jetty-11 when this was the main API for response.    Let's consider this in #11325 for the next release.